### ISSUE TITLE
Emit correct tile_shapes for tensors

### ISSUE THREAD

**Issue by jnie-TT** [tenstorrent/tt-mlir#272]
Currently the tile_shape in the flatbuffer executable is hardcoded to 0x0. In runtime we want to use the tile_shape to determine if we want to tilize/untilize a tensor - some ops (like relu) can only run on tilized data. 

We need compiler support that can emit the correct tile_shapes for input/output tensors.

**Comment by mbezuljTT** [comment]
Hello, 

I had a chat with @derdeljanTT and he reminded me our goal is to maintain cross-platform compiler. Today this means any **linking to TTNN will break cross-platform compilation**. While TTNN might be cross platform in the future, we should be careful about picking this dependency. @nsmithtt  what do you think about it?

Going with **analysis path** means we have to hardcode op requirements manually. 

Last time we have discussed op interface constraints, suggestion from TTNN folks was to build **op constraints cache** and establish straightforward way to rebuilding it. @nsmithtt  are you aware of any development on this front or it's still up for a discussion?

I think we we have a cache of op constraints we could include it in the analysis path - this sounds easier than expanding TTNN.

I see this issue is blocking P0. @jnie-TT  **short-term** is it reasonable to assume all ops can run on tilized data? if yes, are we ok to assume this on the compiler as well? It will come at perf cost. 

Thanks
Marko

**Comment by nsmithtt** [comment]
I think let's shoot for the *op constraints cache* (or config file) approach as that solves the cross platform issue too. @xanderchin, are you using this approach for ttnn sweep tests?

**Comment by mbezuljTT** [comment]
(bumping an old issue)

Did I follow correctly that our default assumption is that we are running row_major/scalar data? Based on what I saw, default assumption should be data is tilized.

If my understanding of default assumption is correct, what was rationale of setting it this way?

**Comment by nsmithtt** [comment]
@mbezuljTT, I think this issue is covering the fact that the compiler is currently not properly modeling the layout of tensors throughout the graph.  We can't really accurately do this without potentially running graph capture to determine what state a particular op has left a tensor in.  At the moment many of these tensor layout attributes that are programmed into the flatbuffer are just incorrect because of this, including tile shape.

> Did I follow correctly that our default assumption is that we are running row_major/scalar data? Based on what I saw, default assumption should be data is tilized.

I don't think there is a default assumption to begin with, compiler doesn't program anything for tile shape because it doesn't know if an op has left it tiled or row major, it depends on the op.  Hence this attribute emitted from compiler is just 0x0.

**Comment by sdjordjevicTT** [comment]
I took a look at the ops that we introduced in the TTNN dialect, and I analyzed their supported layouts. It seems that TILE_LAYOUT support prevails, and I agree with what Marko said, maybe we should consider TILE_LAYOUT to be the default and start modeling compiler default behavior regarding the layout. Of course, we don't live in an ideal world, hence there will be some cases when this isn't true, but I would rather tweak those unique cases in the compiler part of the stack instead of a runtime. Also, in those cases when this isn't true we can treat them as bugs and file metal folks an issue to support the default tile layout. What do you think about this approach?

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Calibri;font-size:14.0pt'>


<div style='direction:ltr'>


Op | Op Type | Suppoted Layout
-- | -- | --
Abs | Elementwise Unary | TL, RM
Sqrt | Elementwise Unary | TL, RM
Neg | Elementwise Unary | TL, RM
Reciprocal | Elementwise Unary | TL, RM
Relu | Elementwise Unary | TL, RM
Sigmoid | Elementwise Unary | TL, RM
Exp | Elementwise Unary | TL, RM
Softmax | Elementwise Unary | TL
Add | Elementwise Binary | TL
Div | Elementwise Binary | TL
Subtract | Elementwise Binary | TL
Multiply | Elementwise Binary | TL
Ge | Elementwise Binary | TL
Maximum | Elementwise Binary | TL
Sum | Reduction | TL
Mean | Reduction | TL
Max | Reduction | TL
Matmul | Matmul | TL
Embedding | Embedding | TL, RM
Transpose | Data Movement | TL, RM
Reshape | Data Movement | TL, RM
Concat | Data Movement | TL, RM
Conv2dOp | Convolution | ?
MaxPool2d | Pool | ?



</div>


</body>

</html>

**Comment by nsmithtt** [comment]
+1 for default to tile layout.

- Conv2d and MaxPool2d need RM inputs, they might run untilize on the fly though

**Comment by sdjordjevicTT** [comment]
We can create a MLIR trait to tag such specific ops in TTNN dialect with something like: SupportOnlyRowMajorLayout

Than, during ttir to ttnn conversion we can convert tensors for such ops from TILE_LAYOUT to ROW_MAJOR_LAYOUY, and leave runtime clean of the hacks(keep running in the runtime only what compiler provided)...

**Comment by nsmithtt** [comment]
I think we already have that with operand constraints, or at least that's what they were intended for.  But we can express it differently if it makes sense.

**Comment by svuckovicTT** [comment]
>I think we already have that with operand constraints, or at least that's what they were intended for. But we can express it differently if it makes sense.

Can we use those constraints when going from forge? We definitely need a mechanism to fail early on this.

**Comment by nsmithtt** [comment]
Yeah agreed, we should move forward with this refactor.  I think the operand constraints are probably in the wrong place because they are intended to enforce backend constraints, they might be better suited to live inside of the backend dialect.

**Comment by skhatriTT** [comment]
@nsmithtt @sdjordjevicTT Is there any update on this issue? The last discussion was about moving operand constraint to backend dialects, wondering if there's been any progress on this?

**Comment by nsmithtt** [comment]
@sdjordjevicTT / @jnie-TT we should be good to close this I think right? Or still an issue?

**Comment by jnie-TT** [comment]
I believe this has been fixed, tile shapes are now correctly determined based on the layout/ElementType in the compiler.

**Comment by sdjordjevicTT** [comment]
Yes, we can close this one, I will close it.
